templates/components/engines/typescript/agent/tools/openapi-action.ts

```diff
@@ -0,0 +1,182 @@
+          description: "The url to make the get request against",
+        },
+        params: {
+          type: "object",
+          description: "the parameters to provide with the get request",
+        },
+      },
+      required: ["url"],
+    },
+  } as const;
+
+  private readonly POST_REQUEST_SPEC = {
+    name: "post_request",
+    description: "Use this to POST content to a website.",
+    parameters: {
+      type: "object",
+      properties: {
+        url: {
+          type: "string",
+          description: "The url to make the get request against",
+        },
+        data: {
+          type: "object",
+          description: "the key-value pairs to provide with the get request",
+        },
+      },
+      required: ["url"],
+    },
+  } as const;
+
+  private readonly PATCH_REQUEST_SPEC = {
+    name: "patch_request",
+    description: "Use this to PATCH content to a website.",
+    parameters: {
+      type: "object",
+      properties: {
+        url: {
+          type: "string",
+          description: "The url to make the get request against",
+        },
+        data: {
+          type: "object",
+          description: "the key-value pairs to provide with the get request",
+        },
+      },
+      required: ["url"],
+    },
+  } as const;
+
+  constructor(
+    public openapi_uri: string,
+    public domainHeaders: DomainHeaders = {},
+  ) {}
+
+  async loadOpenapiSpec(input: { url: string }): Promise<any> {
+    try {
+      const api = (await SwaggerParser.validate(input.url)) as any;
+      return {
+        servers: api.servers,
+        description: api.info.description,
+        endpoints: api.paths,
+      };
+    } catch (err) {
+      return err;
+    }
+  }
+
+  async getRequest(input: { url: string; params: object }): Promise<any> {
+    if (!this.validUrl(input.url)) {
+      return this.INVALID_URL_PROMPT;
+    }
+    try {
+      const res: AxiosResponse = await axios.get(input.url, {
+        headers: this.getHeadersForUrl(input.url),
+        params: input.params,
+      });
+      return res.data;
+    } catch (error) {
+      return error;
+    }
+  }
+
+  async postRequest(input: { url: string; data: object }): Promise<any> {
+    if (!this.validUrl(input.url)) {
+      return this.INVALID_URL_PROMPT;
+    }
+    try {
+      const res: AxiosResponse = await axios.post(input.url, input.data, {
+        headers: this.getHeadersForUrl(input.url),
+      });
+      return res.data;
+    } catch (error) {
+      return error;
+    }
+  }
+
+  async patchRequest(input: { url: string; data: object }): Promise<any> {
+    if (!this.validUrl(input.url)) {
+      return this.INVALID_URL_PROMPT;
+    }
+    try {
+      const res = await axios.patch(input.url, input.data, {
+        headers: this.getHeadersForUrl(input.url),
+      });
+      return res.data;
+    } catch (error) {
+      return error;
+    }
+  }
+
+  public toToolFunctions = () => {
+    return [
+      FunctionTool.from(
```

Comment:
@leehuwuj do you use the same structure for python?

Reply:
Yes, we also have these function tools in python.